lib/Parse/ParseDecl.cpp

```diff
@@ -118,19 +104,21 @@ namespace {
     template <typename T>
     ParserResult<T>
     fixupParserResult(T *D) {
-      if (CC.hasValue()) {
-        swapDecl(D);
+      if (movedToTopLevel()) {
+        D->setHoisted();
+        SF->addHoistedDecl(D);
+        getDebuggerClient()->didGlobalize(D);
```

Comment:
There is an invariant here, I'm guessing that these three always go together. If so, how about factoring this out, and lines 118-121 into a separate function? I'm guessing you've also added assertions for the invariant.